Support Global or Common application config map

shahbour opened the pull request:
Support Global config map called application
Support Specific config map per active profile
So a service with name sample and active profiles production , mysql will try to get the following config maps

application
sample
sample-production
sample-mysql

This is related to [issue](https://github.com/spring-cloud-incubator/spring-cloud-kubernetes/issues/153) create by me

iocanel commented:
I have mixed feelings about this.

I do like the idea of being able to have shared configuration between multiple pods, this is really great.

I am not sure if doing that via the application configmap is the way to go. Why?
- the name is misleading
- it changes the way that application is used so far.

So I'd rather use a different name, like `shared` or `global`. @spencergibb: Your thoughts?

shahbour commented:
@spencergibb @iocanel Any update on this?

spencergibb commented:
I'm fine with a different name as long as it is documented

shahbour commented:
Ok great, I will go with global and update the code

shahbour commented:
I changed the name to global and updated documentation and tests, below are how it works now

Support Global config map called global
Support Specific config map per active profile
So a service with name sample and active profiles production, mysql, will try to get the following config maps

global
sample
sample-production
sample-mysql

geoand commented:
I am working on more generalized solution to this problem that would allow users to specify arbitrary `ConfigMaps` (name and namespace) instead of requiring specific names.
I am hoping to have a working prototype soon (supporting context reload might be a challenge, otherwise it seems doable)

shahbour commented:
After rebasing and getting to latest commit I noticed that if the config map does not have an active profile (spring.profiles) is empty or not set then it is ignored and not added, this was not the case before.

Is this on purpose or it is a bug?

iocanel commented:
@geoand do u have an idea?

geoand commented:
@iocanel I will check in a few hours and let you know

geoand commented:
@iocanel @shahbour The change that was introduced in #174 basically fixes #106 for the 0.3.x release. 
The initial refactoring that was done to bring the project inline with Spring Boot 2.x, had introduced a regression and #106 was once again an issue.

As far as I can tell, the state of the `yaml` handling currently in `master` is the intended one, meaning that it's working as it was for `0.2.0.RELEASE` and not `0.3.0.RC1` which broke the intended functionality.

@shahbour Could you provide an example of what you think might not be working correctly?

shahbour commented:
I will try to do a test unit that fails. To explain what I faced. After I did the update all my config maps that don't include `spring.profiles` were ignored for example the below config map did not work until I added spring.profiles: kubernetes

```
apiVersion: v1
data:
  application.yml: |-
    logging:
      level:
        com.xxxxxx: INFO
    spring:
      profiles: kubernetes
      cache:
        type: hazelcast
```

I think the reason for this is this line 
```
yamlFactory.setDocumentMatchers(
					(DocumentMatcher) properties ->
						(asList(profiles).contains(properties.getProperty("spring.profiles")) ?
							FOUND : NOT_FOUND)
				);
```

geoand commented:
@shahbour I can see why that would be a problem, however the current behavior is the same as existed in `0.2.0.RELEASE`.
If anything I would say that automatically adding the `kubernetes` is problematic. @iocanel Do you remember what the reasoning  was for automatically adding the `kubernetes` profile?

geoand commented:
@shahbour  @iocanel  PR #193 seems like something that would make profile handling feel a little more natural

iocanel commented:
@shahbour @geoand: Is this still relevant?

geoand commented:
I would say that since #176 was merged, this PR becomes obsolete

shahbour commented:
Ok cool, I think we can close this. I will update my code to and try #176 and if anything I will open issue.
